### PR TITLE
Add improved goodness of fit implementation

### PR DESCRIPTION
**Pull request by stes**
This adds a better goodness of fit measure. Instead of the old variant which simply matched the InfoNCE and depends on the batch size, the proposed measure

- is at 0 for "chance level" (vs. log batch size)
- does not need an adjustment for single session vs. multi-session solvers
- increases as the model gets better, which might be more intuitive

The conversion is quite simply done via

```
GoF(model) = log (batch_size_per_session * num_sessions) - InfoNCE(model)
```

This measure is also used in [DeWolf et al., 2024, Eq. (43)](https://www.biorxiv.org/content/10.1101/2024.09.11.612513v1)

image

Application example (GoF improves from 0 to a larger value during training):

image

---

OLD example, for comparison

image

---

Close https://github.com/AdaptiveMotorControlLab/CEBRA-dev/pull/669

### CODE REVIEW

**Comment by stes** [comment]
TODO: Fix case where batch_size is `None`
image

**Comment by CeliaBenquet** [comment]
@stes about what I implemented in #202 that I do see here. 

I think it would be good to have a really basic function where you provide the loss and the batch size, so that it is easily usable in the pytorch implementation as well.

Also, it would be nice to test for the default `CEBRA.batch_size = None`, not sure it is handled here.

File: cebra/integrations/sklearn/metrics.py [hunk 1]
```diff
@@ -108,6 +108,117 @@ def infonce_loss(
...
+
+    Related:
+        :func:`infonce_to_goodness_of_fit`
+
+    Example:
+
+        >>> import cebra
+        >>> import numpy as np
+        >>> neural_data = np.random.uniform(0, 1, (1000, 20))
+        >>> cebra_model = cebra.CEBRA(max_iterations=10)
```
Comment (stes) [inline cebra/integrations/sklearn/metrics.py:141]:
```suggestion
        >>> cebra_model = cebra.CEBRA(max_iterations=10, batch_size = 512)
```

File: cebra/integrations/sklearn/metrics.py [hunk 2]
```diff
@@ -108,6 +108,117 @@ def infonce_loss(
...
+        model: The trained CEBRA model
+
+    Returns:
+        Numpy array containing the goodness of fit values, measured in bits
+
+    Raises:
+        RuntimeError: If the provided model is not fit to data.
+    """
+    if not hasattr(model, "state_dict_"):
+        raise RuntimeError("Fit the CEBRA model first.")
```
Comment (stes) [inline cebra/integrations/sklearn/metrics.py:212]:
```suggestion
    if not hasattr(model, "state_dict_"):
        raise RuntimeError("Fit the CEBRA model first.")
    if model.batch_size is None:
        raise ValueError(
            "Computing the goodness of fit is not yet supported for "
            "models trained on the full dataset (batchsize = None). " 
        )
```

**Comment by stes** [comment]
Unrelated build issue due to upstream change in sklearn (#204 ), attempted fix in #205

**Comment by stes** [comment]
The build issue is fixed, and once #205 is merged tests should pass here as well.

File: cebra/integrations/sklearn/metrics.py [hunk 3]
```diff
@@ -108,6 +108,138 @@ def infonce_loss(
...
+        :func:`infonce_to_goodness_of_fit`
+
+    Example:
+
+        >>> import cebra
+        >>> import numpy as np
+        >>> neural_data = np.random.uniform(0, 1, (1000, 20))
+        >>> cebra_model = cebra.CEBRA(max_iterations=10, batch_size = 512)
+        >>> cebra_model.fit(neural_data)
+        CEBRA(max_iterations=10)
```
Comment (stes) [inline cebra/integrations/sklearn/metrics.py:174]:
```suggestion
        CEBRA(batch_size=512, max_iterations=10)
```

**Review by CeliaBenquet — CHANGES_REQUESTED** [review]
Thank you @stes! This looks nice!! Some minor suggestions on the docstrings and maybe add some tests for the different corner cases based on the arguments provided in `infonce_to_goodness_of_fit`.

File: tests/test_sklearn_metrics.py [hunk 4]
```diff
@@ -383,3 +383,67 @@ def test_sklearn_runs_consistency():
...
+
+    history_random = _fit_and_get_history(X, y_random)
+    history_linear = _fit_and_get_history(X, y_linear)
+
+    assert isinstance(history_random, np.ndarray)
+    assert history_random.shape[0] > 0
+    # NOTE(stes): Ignore the first 5 iterations, they can have nonsensical values
+    # due to numerical issues.
+    history_random_non_negative = history_random[history_random >= 0]
+    np.testing.assert_allclose(history_random_non_negative, 0, atol=0.05)
```
Comment (stes) [inline tests/test_sklearn_metrics.py:444]:
```suggestion
    np.testing.assert_allclose(history_random_non_negative, 0, atol=0.075)
```

slightly less conservative threshold due to failed test

**Comment by stes** [comment]
one review comment to resolve (= add more relevant tests for one batchsize None), then ready!

**Review by MMathisLab — APPROVED** [review]
lgtm! thanks
